blockcypher/src/endpoint/balance.ts

```diff
@@ -28,25 +29,86 @@ const getChainId = (coin: CoinType, chain: ChainType): string => {
   }
 }
 
-const getBalance: balance.GetBalance = async (account, config) => {
-  const chainId = getChainId(account.coin as CoinType, account.chain as ChainType)
-  const api = new bcypher(account.coin, chainId, config.apiKey)
+/**
+ * Chunk an array into a nested array
+ *
+ * @param amount number the max size of the chunks
+ * @param data array of arbitrary data
+ *
+ * @returns 2d array
+ */
+const chunk = (amount: number, data: any[]) => {
+  const output: any[][] = []
+  const length = data.length
+  if (amount < 1 || length < 1) return output
+  const chunks = Math.ceil(data.length / amount)
+  for (let i = 0; i < chunks; i++) {
+    const offset = amount * i
+    const slice = data.slice(offset, offset + amount)
+    output.push(slice)
+  }
+  return output
+}
+
+/**
+ * Delays the amount of requests sent per second
+ *
+ * @param amount maximum number of requests per second
+ * @param data array of arbitrary data
+ * @param exec function handler of a chunk
+ *
+ * @returns array of response data
+ */
+const throttle = async (amount: number, data: any[], exec: any) => {
```

Comment:
Rate limiting, not throttle is the concept we are looking for.

Research rate limit algorithms:

- https://blog.logrocket.com/rate-limiting-node-js/
- https://kendru.github.io/javascript/2018/12/28/rate-limiting-in-javascript-with-a-token-bucket/
- https://aliz.ai/rate-limiting-in-rxjs/

Pick one and implement it to be used like this:

```javascript
const count = 5
const slidingWindowTime = 1000
const _echo = (msg) => console.log(msg)
const _limitedEcho = utils.rateLimit(count, slidingWindowTime, _testFn)

while (true) _limitedEcho("Hello!") // This should write 5x "Hello!", then sleep until the end of the sliding window, and repeat indefinitely
```

Please also document the algorithm you picked and implemented.

Reply:
Thanks for pointing me in the right direction with the name :sweat_smile:
I renamed this to `rate_limit`.

How do you feel about moving this to a separate issue? PoR is a high priority item for the business side and I worry making this change might take a little bit to implement. I know @RodrigoAD is working on a similar thing, he could take this over to make it very re-usable.